### PR TITLE
SCP 3192: ACTUS module refactoring

### PR DESCRIPTION
**Pull request by yveshauser**
This PR bundles a couple of (unrelated) changes
* Introduction of a separate `actus-core` package that does not depend on Cardano libraries
* Introduction of a `actus-core-test` with the ACTUS test framework
* New ACTUS contract types
    * CEG: Guarantee
    * CEC: Collateral
* Preserving financial rounding for ACTUS contracts

Pre-submit checklist:
- Branch
    - [x] Tests are provided (if possible)
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
    - [x] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested

### CODE REVIEW

File: marlowe-cli/examples/actus/ReadMe.md [hunk 1]
```diff
@@ -0,0 +1,804 @@
...
+
+Because this is a role-based contract, we compute the address of the script for roles.
+
+```
+ROLE_ADDRESS=$(jq -r '.rolesValidator.address' tx-1.marlowe)
+```
+
+The role address is `addr_test1wr6m6dv8v75gl7hfmnzaltlrph8wr90vqg0m72a4gwrzphcd7qsj4`.
+
+The lender John Fletcher submits the transaction along with the minimum ADA 3000000 lovelace required for the contract's initial state. Submitting with the `--print-stats` switch reveals the network fee for the contract, the size of the transaction, and the execution requirements, relative to the protocol limits.
```
Comment (palas) [inline marlowe-cli/examples/actus/ReadMe.md:349]:
```suggestion
The lender John Fletcher submits the transaction along with the minimum ADA amount of 3,000,000 lovelace required for the contract's initial state. Submitting with the `--print-stats` switch reveals the network fee for the contract, the size of the transaction, and the execution requirements, relative to the protocol limits.
```

File: marlowe-cli/examples/actus/ReadMe.md [hunk 2]
```diff
@@ -0,0 +1,804 @@
...
+
+```console
+Fee: Lovelace 191241
+Size: 572 / 32768 = 1%
+Execution units:
+  Memory: 0 / 30000000 = 0%
+  Steps: 0 / 10000000000 = 0%
+```
+
+The contract received the minimum ADA of 3000000 lovelace from the lender John Fletcher in the transaction `06ee9291aada5c6d8a6f4a29646f5c0d4619697a5f9ccd0de9313256f02635f1`.  Here is the UTxO at the contract address:
```
Comment (palas) [inline marlowe-cli/examples/actus/ReadMe.md:374]:
```suggestion
The contract received the minimum ADA amount of 3,000,000 lovelace from the lender John Fletcher in the transaction `06ee9291aada5c6d8a6f4a29646f5c0d4619697a5f9ccd0de9313256f02635f1`.  Here is the UTxO at the contract address:
```

File: marlowe-cli/examples/actus/ReadMe.md [hunk 3]
```diff
@@ -0,0 +1,804 @@
...
+
+```console
+Fee: Lovelace 1397789
+Size: 13750 / 32768 = 41%
+Execution units:
+  Memory: 7501806 / 30000000 = 25%
+  Steps: 2564286091 / 10000000000 = 25%
+```
+
+The contract passed the deposit of 10000 ADA in the transaction `ddbfdb164c204f627f8ab7f849c96a901901d0b6fecb2307eda60ca1f921d5ef` from the lender to the role address, for the benefit of the borrower. Here is the UTxO at the contract address:
```
Comment (palas) [inline marlowe-cli/examples/actus/ReadMe.md:451]:
```suggestion
The contract passed the deposit of 10,000 ADA in the transaction `ddbfdb164c204f627f8ab7f849c96a901901d0b6fecb2307eda60ca1f921d5ef` from the lender to the role address, for the benefit of the borrower. Here is the UTxO at the contract address:
```

File: marlowe-cli/examples/actus/ReadMe.md [hunk 4]
```diff
@@ -0,0 +1,804 @@
...
+
+```console
+Fee: Lovelace 1112645
+Size: 13246 / 32768 = 40%
+Execution units:
+  Memory: 4300844 / 30000000 = 14%
+  Steps: 1478686295 / 10000000000 = 14%
+```
+
+The closing of the contract paid in the transaction `20f1986772267260e2c3ef4dd5f5c78b94ab41b97385dff8452818860bdf47cd` the 10000 ada principal and 200 ada interest to the role address for the benefit of the lender John Fletcher, along with the minimum ADA 3000000 lovelace that they deposited when creating the contract. There is no UTxO at the contract address:
```
Comment (palas) [inline marlowe-cli/examples/actus/ReadMe.md:655]:
```suggestion
The closing of the contract paid in the transaction `20f1986772267260e2c3ef4dd5f5c78b94ab41b97385dff8452818860bdf47cd` the 10,000 ADA principal and 200 ADA interest to the role address for the benefit of the lender John Fletcher, along with the minimum ADA amount of 3,000,000 lovelace that they deposited when creating the contract. There is no UTxO at the contract address:
```

File: marlowe-cli/examples/actus/ReadMe.md [hunk 5]
```diff
@@ -0,0 +1,804 @@
...
+  "statusDate": "$STATUS_DATE",
+  "cycleOfInterestPayment": "P1YL1",
+  "prepaymentEffect": "N",
+  "nominalInterestRate": `jq -n $INTEREST_RATE`,
+  "interestCalculationBase": "NT"
+}
+EOF
+```
+
+The contract has a minimum-ADA requirement and two timeouts. It also specifies that the lender John Fletcher will pay principal of 10000 ada before Fri, 12 Aug 2022 08:11:01 +0000 and the borrower will repay the principal and interest of 200 ada before Fri, 12 Aug 2022 20:11:01 +0000.
```
Comment (palas) [inline marlowe-cli/examples/actus/ReadMe.md:301]:
```suggestion
The contract has a minimum-ADA requirement and two timeouts. It also specifies that the lender John Fletcher will pay principal of 10,000 ADA before Fri, 12 Aug 2022 08:11:01 +0000 and the borrower will repay the principal and interest of 200 ADA before Fri, 12 Aug 2022 20:11:01 +0000.
```

**Review by palas — APPROVED** [review]
Looks good from what I can see. I just pointed out a couple of details in case they help

File: marlowe-actus/actus-core/src/Actus/Model/StateTransition.hs [hunk 6]
```diff
@@ -0,0 +1,1094 @@
...
+  s
+  t =
+    let timeFromLastEvent = yearFraction dcc (s ^. L.statusDate) t maturityDate
+        timeFromLatestFeePayment = yearFraction dcc latestFeePayment t maturityDate
+        timeFromLatestToNextFeePayment = yearFraction dcc latestFeePayment nextFeePayment maturityDate
+        delta_r = _min (_max (o_rf_RRMO * rrmlt + rrsp - (s ^. L.nominalInterest)) rrpf) rrpc
+        ipnr' = _min (_max (s ^. L.nominalInterest + delta_r) rrlf) rrlc
+     in _STF_PY_PAM fs ct s t
+          & L.accruedInterest .~ (s ^. L.accruedInterest) + timeFromLastEvent * (s ^. L.nominalInterest) * (s ^. L.notionalPrincipal)
+          & L.accruedFees .~ (if latestFeePayment == nextFeePayment then 0 else (timeFromLatestFeePayment / timeFromLatestToNextFeePayment) * sign contractRole * fer)
```
[the remark this answers is not in the document]
Comment (yveshauser) [reply marlowe-actus/actus-core/src/Actus/Model/StateTransition.hs:673]: There is no longer the `Eq` constraint on the value and as a workaround we just compare the `LocalTime`s. 
The tests in the ACTUS test-framework use `Double` and therefore we have to deal with unsafe division.

File: marlowe-cli/examples/actus/ReadMe.md [hunk 7]
```diff
@@ -0,0 +1,804 @@
...
+### Select Network
+
+```
+if [[ -z "$MAGIC" ]]
+then
+  MAGIC=1567
+fi
+```
+
+MAGIC=1566
```
[the remark this answers is not in the document]
Comment (yveshauser) [reply marlowe-cli/examples/actus/ReadMe.md:29]: The file `ReadMe.md` is generated by `make ReadMe.md` - the `MAGIC` is put into the file for reference:
https://github.com/input-output-hk/marlowe-cardano/blob/b1d8a38767365185dc360ccaf669b0eaf99096b1/marlowe-cli/examples/actus/run-actus.sh#L31

File: marlowe-cli/examples/actus/ReadMe.md (hunk 1 above)
Comment (yveshauser) [reply marlowe-cli/examples/actus/ReadMe.md:356]: Thanks, added `amount of` here

**Pull request closed without merging**